### ISSUE TITLE
Track the last time when the entity was changed

### ISSUE THREAD

**Issue by iriswang233** [backstage/backstage#14181]
## Feature Suggestion

Track the last time when the entity was changed

## Possible Implementation

Maybe a new db field like ```last_changed_at``` can be added in refresh_state database table. Once an entity has been discovered, maybe based on  whether the ``` result_hash```  in ```refresh_state``` table is changed, the ```last_changed_at``` can be updated accordingly.

## Context
This will help track the freshness of the entity. If an entity was not updated for 1 month, it can be considered as outdated. We are trying to improve the entity update process like notifying entity owners if they haven't update their entity for long time.

**Comment by varadkarpe** [comment]
Hey @iriswang233 
What do you suggest we do with the existing entities which don't have `last-changed-at` attribute?

**Comment by kissmikijr** [comment]
Hm, interesting. I'd assume entities are achieving a final(ish) state? When they are not really changing that much. Why do you think that an entity should change all the time ?

**Comment by iriswang233** [comment]
> Hey @iriswang233 What do you suggest we do with the existing entities which don't have `last-changed-at` attribute?

Hey @varadkarpe varadkarpe,

Good point!

For existing entities which doesn't have `last_changed_at`, maybe there is a need to add `last_changed_at` with value same with `last_discovery_at` or empty value for all of them.  This can be a batch update.

Or instead of a new attribute as a new column in DB, maybe it's possible to add it as an attribute in one of existing columns maybe `processed_entity` in the `refresh_state` table? For example,

```
{
  "metadata": {...},
  "apiVersion": "backstage.io/v1alpha1",
  "kind": "Location",
  "spec": {... },
"last_changed_at": xxx
}
```
Maybe there is better solution? Please advice.

**Comment by iriswang233** [comment]
hi @kissmikijr ,

Yes, I agree with you that ideally entity won't change often and they should be maintained responsively by owner. 

However, out problem is many entities were created with insufficient information or became outdated due to various reasons, like owner didn't have enough information when they created the entity and forgot to update later, there is no serious consequence if the entity is outdated so owners deprioritised the update and forgot later, 

And, our owners are still adopting backstage product and concept so there are many incorrect information in entities. Now we have higher adoption rate so we wanna improve the entity quality through notifying owners to review their entities after certain period instead of forgetting them after creation.

I hope i explain the context better. Please let me know if you need more clarification.

**Comment by andrewthauer** [comment]
> Is it worth it? What do you think @backstage/catalog-core?

I was recently looking for something like this the other day. Been doing some data dumps from various internal systems/sources, one being the catalog. Having this available either in the final_entity as an annotation or a db column would be useful in doing some auditing/control checking.

**Comment by freben** [comment]
The original ask was for the _incoming_ (raw, unprocessed) entity. Correct me if I misread. The outgoing (stitched) entity could get updated any number of times for unrelated reasons - for example, if some _other_ entity emits a relation that points to an entity, it'll get restitched.

I would argue that that data should not necessarily go in the catalog at all; at least, it should not go into the entity body. If you wanted, you could make an external process that lists all entities, finds by annotation the URL that they are sourced from, and ask the authority (the source control system itself) when the file was last updated.

But yeah, I do also agree with the sentiment that the proposed process sounds like it would unnecessarily bother those that have old, stable, long running things whose metadata have no reason to change.

**Comment by andrewthauer** [comment]
> and ask the authority (the source control system itself) when the file was last updated.

So this makes sense conceptually. The authority of the data makes sense. However, what about cases where the entity is enriched via `CatalogProcessors` or even placeholders? Only relying on the catalog-info.yaml changes would not factor in data that the catalog file points to right?

**Comment by freben** [comment]
Yeah that's right. So it will depend on what one is looking for. It may be the case that there are a number of these timestamps with different semantics that could be interesting.

- last time we tried to look at the source data
- last time we observed a change in the source data
- last time the unprocessed entity changed (remember, the source data may be anything, and can contain many entities)
- last time the processed entity changed
- last time the final entity changed

and all of those could have etag-like info - some of them already have that, in the form of the hashes.

**Comment by iriswang233** [comment]
The change made by custom processors also matters to us, these 2 should be related to my request.
- last time the processed entity changed
- last time the final entity changed

**Comment by iriswang233** [comment]
hi  @https://github.com/orgs/backstage/teams/catalog-core
any update on this request?

**Comment by iriswang233** [comment]
hi team,
Can i contribute to it? i will need some guidance.

For last time the final entity changed, it could be added like this, right? https://github.com/backstage/backstage/issues/14181#issuecomment-1285404139

How about last time the processed entity changed?

**Comment by freben** [comment]
I'm arguing that instead, we should first make sure to track this in the catalog database tables. Putting (especially time based) volatile information in the catalog entities is probably unfortunate. Either way, after we have this info in the database, we could consider reading it up and adding it to entities as part of the stitching step later on.

Main task is to add columns to the proper tables (see the `migrations` directory), and then to make sure to write to those columns in the required places (see especially the `DefaultEntitiesCatalog` and `Stitcher` classes).

**Comment by iriswang233** [comment]
could you help review https://github.com/backstage/backstage/pull/14980?

**Comment by iriswang233** [comment]
hi team, 
last_updated_at has been added in final_entities table in this PR https://github.com/backstage/backstage/pull/14980, what would be a good  way to expose it to entity catalog API?

**Comment by freben** [comment]
Pinging @backstage/catalog-core for this one.

One possible suggestion might be to mimic the kubernetes model, [where they add timestamps](https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values) directly under `metadata`, and reserving those field names as well-known in our docs and in the core entity model. We should then reuse the names `creationTimestamp` / `deletionTimestamp` whenever they become relevant in the future, and pick e.g. `updatedTimestamp` (don't think there's precedent set here?) for the timestamp at hand here.

**Comment by alde** [comment]
I think adding it under metadata, as @freben said, makes sense - especially since there is prior art in the inspiration for the catalog model. Naming things is hard, but I'd be fine with `updatedTimestamp` or `modifiedTimestamp` or something along those lines.

**Comment by brianphillips** [comment]
The way I've been thinking about this is the "last updated" date is less interesting than something along the lines of "last confirmed up-to-date" date. Things can get updated all the time for various reasons but that doesn't say anything about whether the full set of data is up to date. Similarly, just because details about a catalog entity have _not_ been changed recently, doesn't mean it's no longer accurate.

In view of this, I'd like a `metadata` field that can be set to a specific date (by a human or other process that has knowledge of the accuracy of the data about the entity). That could then be used to drive entity warnings (i.e. "This entity has not been reviewed for accuracy since <some date>") or Tech Insights or Soundcheck compliance report based on some sort of configured tolerance for entity accuracy review.

Thoughts?

**Comment by Rugvip** [comment]
@brianphillips we're hesitant to add any kind for forced manual updates to the common catalog schema at this point, but you can always [extend your own catalog model](https://backstage.io/docs/features/software-catalog/extending-the-model) to support it.

**Comment by khanagan** [comment]
any update on this?
